torch/autograd/profiler.py

```diff
@@ -366,14 +366,40 @@ class record_function(ContextDecorator):
     """
     def __init__(self, name):
         self.name = name
+        # Whether or not we should run record function's end callbacks when exiting.
+        self.run_callbacks_on_exit = True
 
     def __enter__(self):
         self.handle = torch.ops.profiler._record_function_enter(self.name)
+        return self
 
     def __exit__(self, *args):
-        torch.ops.profiler._record_function_exit(self.handle)
+        if self.run_callbacks_on_exit:
+            torch.ops.profiler._record_function_exit(self.handle)
         return False
 
+    def _call_end_callbacks_on_future(self, fut):
+        """
+        _call_end_callbacks_on_future is meant to be used for profiling async
```

Comment:
Can we extend the docs here mentioning that this can be called only once? Also, lets add an assertion ensuring that if the user calls this twice they get an appropriate error the second time.

Reply:
Are we sure we want to throw an error, or should the 2nd call just be a no-op?

Reply:
i'd be more inclined towards throwing an error, this shouldn't normally happen